### PR TITLE
[#440, #487] Conversion of view types and automatic replacing of creatable views by context dependent declared view type

### PR DESCRIPTION
**Pull request by beikov**
## Related Issue

#440 and #487

### CODE REVIEW

**Comment by Mobe91** [comment]
IMO, you are missing a test case that tests conversion of subview mappings. E.g.

```java
@EntityView(Document.class)
public interface DocumentIdView extends Serializable {

    @IdMapping
    public Long getId();

    public SimplePersonView getPeople();

}
@EntityView(Document.class)
public interface DocumentCloneView extends DocumentIdView {
    public PersonView getPeople();
}
```

When you convert from DocumentCloneView to DocumentIdView, then also the people property should be converted from PersonView to SimplePersonView, right?

File: entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/mapper/ViewMapper.java [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+        private final ManagedViewType<T> targetType;
+        private final boolean ignoreMissing;
+
+        public Key(ManagedViewType<S> sourceType, ManagedViewType<T> targetType, boolean ignoreMissing) {
+            this.sourceType = sourceType;
+            this.targetType = targetType;
+            this.ignoreMissing = ignoreMissing;
+        }
+
+        public ViewMapper<S, T> createMapper(ProxyFactory proxyFactory) {
```
Comment (Mobe91) [inline entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/mapper/ViewMapper.java]: Why is the `Key` able to create a `ViewMapper`? Shouldn't the `Key` be merely a value holder to facilitate a compound map key? Looks to me like violating the single responsibility principle for no reason.

File: entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/update/flush/CompositeAttributeFlusher.java [hunk 2]
```diff
@@ -350,40 +360,80 @@ public boolean flushEntity(UpdateContext context, Object entity, Object view, Ob
...
+                                removedValue = recordingMap.getDelegate().remove(updatableProxy);
+                            }
                         } else {
-                            removedValue = recordingMap.getDelegate().remove(updatableProxy);
+                            if (removedKeys == null) {
+                                removedKeys = new HashSet<>();
+                            }
+                            // TODO: replaceValue currently only handles the current value, which is inconsistent regarding what we do in the elementFlusher case
+                            // Not sure if a creatable view should be allowed to occur multiple times in the map as value..
+                            if (recordingMap.getCurrentIterator() != null) {
```
Comment (Mobe91) [inline entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/update/flush/CompositeAttributeFlusher.java]: Positive logic

File: entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/update/flush/CompositeAttributeFlusher.java [hunk 3]
```diff
@@ -350,40 +360,80 @@ public boolean flushEntity(UpdateContext context, Object entity, Object view, Ob
...
+                        recordingCollection.getCurrentIterator().add(newObject);
                     } else {
                         recordingCollection.getDelegate().add(updatableProxy);
                     }
-                } else if (recordingMap != null && recordingMap.isHashBased()) {
-                    if (recordingMap.getCurrentIterator() != null) {
-                        recordingMap.getCurrentIterator().add(updatableProxy, removedValue);
+                } else if (recordingMap != null && (persistViewMapper != null || updatableProxy.$$_getParentIndex() == 1 && recordingMap.isHashBased())) {
+                    if (updatableProxy.$$_getParentIndex() == 1) {
+                        if (recordingMap.getCurrentIterator() != null) {
```
Comment (Mobe91) [inline entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/update/flush/CompositeAttributeFlusher.java]: Positive logic

File: entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/update/flush/CompositeAttributeFlusher.java [hunk 4]
```diff
@@ -350,40 +360,80 @@ public boolean flushEntity(UpdateContext context, Object entity, Object view, Ob
...
+                            recordingMap.getCurrentIterator().add(newObject, removedValue);
+                        } else {
+                            recordingMap.getDelegate().put(newObject, removedValue);
+                        }
                     } else {
-                        recordingMap.getDelegate().put(updatableProxy, removedValue);
+                        for (Object removedKey : removedKeys) {
+                            if (recordingMap.getCurrentIterator() != null) {
```
Comment (Mobe91) [inline entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/update/flush/CompositeAttributeFlusher.java]: Positive logic

**Comment by beikov** [comment]
Merged by ff10bf28802a1642342f5e3a5890d5db06e21224

**Pull request closed without merging**